src/public.rs

```diff
@@ -0,0 +1,283 @@
+
+            (&mut public_with_flag[1..]).copy_from_slice(public.as_ref());
+
+            // Safety: FFI call
+            let curve = ffi::secp256k1_ec_pubkey_parse(
+                ffi::secp256k1_context_no_precomp,
+                &mut ffi::PublicKey::new(),
+                public_with_flag.as_c_ptr(),
+                UNCOMPRESSED_PUBLIC_KEY_SIZE,
+            );
+
+            curve == 1
+        }
+
+        /// Check if the secret key representation is in the curve.
+        pub fn is_in_curve(&self) -> bool {
+            // Safety: struct is guaranteed to reference itself with correct len
+            unsafe { Self::is_slice_in_curve_unchecked(self.as_ref()) }
+        }
+
+        /// Recover the public key from a signature performed with
+        /// [`SecretKey::sign`]
+        pub fn recover<M>(signature: Signature, message: M) -> Result<PublicKey, Error>
+        where
+            M: AsRef<[u8]>,
+        {
+            let message = SecretKey::normalize_message(message);
+
+            Self::_recover(signature, &message)
+        }
+
+        /// Recover the public key from a signature performed with
+        /// [`SecretKey::sign`]
+        ///
+        /// # Safety
+        ///
+        /// The protocol expects the message to be the result of a hash -
+        /// otherwise, its verification is malleable. The output of the
+        /// hash must be 32 bytes.
+        ///
+        /// The unsafe directive of this function is related only to the message
+        /// input. It might fail if the signature is inconsistent.
+        pub unsafe fn recover_unchecked<M>(
+            signature: Signature,
+            message: M,
+        ) -> Result<PublicKey, Error>
+        where
+            M: AsRef<[u8]>,
+        {
+            let message = SecretKey::cast_message(message.as_ref());
+
+            Self::_recover(signature, message)
+        }
+
+        fn _recover(mut signature: Signature, message: &Message) -> Result<PublicKey, Error> {
+            let v = ((signature.as_mut()[32] & 0x90) >> 7) as i32;
+            signature.as_mut()[32] &= 0x7f;
+
+            let v = RecoveryId::from_i32(v)?;
+            let signature = RecoverableSignature::from_compact(signature.as_ref(), v)?;
+
+            let pk = Secp256k1::new()
+                .recover(message, &signature)?
+                .serialize_uncompressed();
+
+            // Ignore the first byte of the compressed flag
+            let pk = &pk[1..];
+
+            // Safety: secp256k1 protocol specifies 65 bytes output
+            let pk = unsafe { Bytes64::from_slice_unchecked(pk) };
+
+            Ok(Self(pk))
+        }
+    }
+
+    impl TryFrom<Bytes64> for PublicKey {
+        type Error = Error;
+
+        fn try_from(b: Bytes64) -> Result<Self, Self::Error> {
+            let public = PublicKey(b);
+
+            public
+                .is_in_curve()
+                .then(|| public)
+                .ok_or_else(|| Secp256k1Error::InvalidPublicKey.into())
```

Comment:
Why not just use our own error types here instead of immediately converting from Secp256k1Error?

Reply:
https://github.com/FuelLabs/fuel-crypto/pull/1/commits/0e3c899963f9e01dc78e5d4499868172264611ff